tee-worker/litentry/core/data-providers/src/nodereal.rs

```diff
@@ -0,0 +1,204 @@
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use crate::sgx_reexport_prelude::chrono::{offset::Utc as TzUtc, DateTime, NaiveDateTime};
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use std::time::{SystemTime, UNIX_EPOCH};
+
+#[cfg(feature = "std")]
+use chrono::{offset::Utc as TzUtc, DateTime, NaiveDateTime};
+
+use crate::{build_client, DataProviderConfig, Error, HttpError};
+use http::header::CONNECTION;
+use http_req::response::Headers;
+use itc_rest_client::{
+	http_client::{DefaultSend, HttpClient},
+	rest_client::RestClient,
+	RestPath, RestPost,
+};
+use litentry_primitives::ErrorDetail;
+use serde::{Deserialize, Serialize};
+use std::{
+	format, str,
+	string::{String, ToString},
+	vec::Vec,
+};
+
+fn now() -> DateTime<TzUtc> {
+	#[cfg(feature = "std")]
+	{
+		TzUtc::now()
+	}
+
+	#[cfg(all(not(feature = "std"), feature = "sgx"))]
+	{
+		let now = SystemTime::now()
+			.duration_since(UNIX_EPOCH)
+			.expect("system time before Unix epoch");
+		let naive =
+			NaiveDateTime::from_timestamp_opt(now.as_secs() as i64, now.subsec_nanos() as u32)
+				.unwrap();
+
+		DateTime::from_utc(naive, TzUtc)
+	}
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct NoderealServiceReqPath {
+	path: String,
+}
+
+impl Default for NoderealServiceReqPath {
+	fn default() -> Self {
+		Self { path: "".to_string() + "/spaceid/domain/names/" }
+	}
+}
+
+impl NoderealServiceReqPath {
+	pub fn new(api_key: &str, tag: &str) -> Self {
+		Self { path: api_key.to_string() + "/spaceid/domain/names/" + tag }
+	}
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct SpaceIDReqBody(Vec<String>);
+impl RestPath<NoderealServiceReqPath> for SpaceIDReqBody {
+	fn get_path(req: NoderealServiceReqPath) -> core::result::Result<String, HttpError> {
+		Ok(req.path)
+	}
+}
+
+pub struct NoderealClient {
+	pub api_key: String,
+	pub api_url: String,
+	client: RestClient<HttpClient<DefaultSend>>,
+}
+
+impl NoderealClient {
+	pub fn new(data_provider_config: &DataProviderConfig) -> Self {
+		let api_key = data_provider_config.nodereal_api_key.to_string();
+		let api_url = data_provider_config.nodereal_api_url.to_string();
+
+		let mut headers = Headers::new();
+		headers.insert(CONNECTION.as_str(), "close");
+		let client = build_client(api_url.as_str(), headers);
+
+		NoderealClient { api_key, api_url, client }
+	}
+}
+
+pub trait NoderealHttpMethos {
+	fn post(
+		&mut self,
+		path: NoderealServiceReqPath,
+		body: &SpaceIDReqBody,
+	) -> Result<serde_json::Value, Error>;
+}
+
+impl NoderealHttpMethos for NoderealClient {
+	fn post(
+		&mut self,
+		path: NoderealServiceReqPath,
+		body: &SpaceIDReqBody,
+	) -> Result<serde_json::Value, Error> {
+		let response = self
+			.client
+			.post_capture::<NoderealServiceReqPath, SpaceIDReqBody, serde_json::Value>(path, body);
+		response.map_err(|e| Error::AchainableError(format!("Nodereal response error: {}", e)))
```

Comment:
is the client also refers to Achainable provider? or Achainable and Nodereal are different providers? I was thinking that if they are different and we want to share common error, then the name might mislead and it may deserve to be named like `DataProviderClientError` or something similar, what do you think?

Reply:
Thanks, fixed now ,this is a copy-paste error...